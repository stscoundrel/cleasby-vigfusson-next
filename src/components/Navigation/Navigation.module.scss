.section {
  color: white;
  background-color: $color-blue;
  padding: 16px 0;
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  z-index: 1989;
  border-bottom: 1px solid white;
}

.topBarSpacer {
  display: block;
  height: 70px;
}

.topbar {
  display: flex;
  justify-content: space-between;
  align-items: center;
  text-align: right;
  height: 38px;
}

.content {
  display: none;
  flex-wrap: wrap;
  justify-content: flex-start;
}

.opened {
  display: flex;
  padding: 30px 0;
}

.list {
  display: flex;
  flex-wrap: wrap;
  flex-basis: 100%;
  justify-content: space-between;
  margin-bottom: 30px;
}

.listItem {
  flex-basis: 15%;
  margin-bottom: 10px;
  text-align: center;
}

@media only screen and (min-width: 768px) {
  .listItem {
    flex-basis: 10%;
    text-align: left;
  }
}

@media only screen and (min-width: 1200px) {
  .section {
    position: initial;
    top: initial;
    left: initial;
    right: initial;
    z-index: initial;
    border-bottom: initial;
  }
  
  .topBarSpacer {
    display: none;
  }

  .content {
    flex-wrap: nowrap;
    padding: 0;
  }

  .list {
    flex-basis: 80%;
    margin-bottom: 0;
    text-align: center;
  }

  .listItem {
    flex: 1;
    margin-bottom: 0;
  }

  .topbar {
    display: none;
  }

  .content {
    display: flex;
  }
}