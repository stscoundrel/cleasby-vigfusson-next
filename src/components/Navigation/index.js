import { useState } from 'react'
import Link from 'next/link'

// Components.
import Hamburger from 'components/Hamburger'
import LetterLink from 'components/LetterLink'
import Search from 'components/Search'

// Styles.
import styles from './Navigation.module.scss'

export default function Navigation({ letters, noSearch = false }) {
  const [isOpen, setIsOpen] = useState(false)

  const openNav = () => {
    setIsOpen(!isOpen)
  }

  const getOpenClass = () => (isOpen ? styles.opened : null)

  return (
    <>
      <div className={styles.topBarSpacer}></div>
      <nav className={styles.section}>
        <div className={`${styles.topbar} container`}>
          <Link href="/">
            <img src="/favicon-48x48.png" width="30" height="30" alt="To home" loading="lazy" />
          </Link>
          <Hamburger action={openNav} />
        </div>
        <div className={`${styles.content} ${getOpenClass()} container`}>
          <ul className={styles.list}>
            {letters.map((entry) => (
              <li className={styles.listItem} key={entry.slug}>
                <LetterLink letter={entry} />
              </li>
            ))}
          </ul>
          { !noSearch && <Search /> }
        </div>
      </nav>
    </>
  )
}
