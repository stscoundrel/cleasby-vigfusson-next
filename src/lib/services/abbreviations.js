import { findAbbreviations, findWorksAndAuthors } from 'cleasby-vigfusson-abbreviations'
import { abbreviate } from 'abbreviatrix'

/**
 * Combine abbreviations from definitions array.
 * Accepts finder method to use.
 */
const combineAbbreviations = ({ definitions }, find) => {
  const combinedAbbreviations = []
  const abbreviationSet = new Set()

  definitions.forEach((definition) => {
    const abbreviations = find(definition)
    abbreviations.forEach((explanation, abbreviation) => {
      if (!abbreviationSet.has(abbreviation)) {
        abbreviationSet.add(abbreviation)
        combinedAbbreviations.push({ abbreviation, explanation })
      }
    })
  })

  return combinedAbbreviations;
}

export const getAbbreviations = (entry) => ({
  common: combineAbbreviations(entry, findAbbreviations),
  works: combineAbbreviations(entry, findWorksAndAuthors),
})

/**
 * Add abbr tags to content with explanations.
 */
export const addAbbreviationsToContent = (content, abbreviations) => {
  const combinedAbbrs = [...abbreviations.common, ...abbreviations.works]
  let result = content

  combinedAbbrs.forEach(({ abbreviation, explanation }) => {
    result = abbreviate(abbreviation, explanation, result)
  })

  return result
}

export default {
  getAbbreviations,
  addAbbreviationsToContent,
}
