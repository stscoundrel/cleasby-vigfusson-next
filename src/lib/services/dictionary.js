const { getDictionary } = require('cleasby-vigfusson-dictionary')
const { VALID_AS_FIRST } = require('old-norse-alphabet')
const { slugifyWord, slugifyLetter } = require('../utils/slugs')

let cachedDictionary = null

const addSlugs = (words) => {
  const existingSlugs = {}

  const formattedWords = words.map((word) => {
    let slug = slugifyWord(word.word).toLowerCase()

    if (slug.charAt(0) === '-') {
      slug = slug.substring(1)
    }

    if (existingSlugs[slug]) {
      // Double slug, make unique.
      existingSlugs[slug] += 1
      slug = `${slug}-${existingSlugs[slug]}`
    } else {
      existingSlugs[slug] = 1
    }

    return {
      ...word,
      slug,
    }
  })

  return formattedWords
}

export const getAllWords = () => {
  if (cachedDictionary) return cachedDictionary

  const words = getDictionary()

  /**
   * Add URL safe slugs.
   */
  const formattedWords = addSlugs(words)

  cachedDictionary = formattedWords

  return formattedWords
}

export const getByLetter = (letter) => {
  const words = getAllWords()
  const byLetter = words.filter((entry) => (
    entry.word.charAt(0).toLowerCase() === letter.toLowerCase()))

  return byLetter
}

export const getWord = (word) => getAllWords().filter((entry) => entry.slug === word)[0]

export const getAlphabet = () => {
  const letters = [...VALID_AS_FIRST.filter((letter) => letter !== 'ǫ' && letter !== 'ø'), 'ö']

  const formattedLetters = letters.map((letter) => ({
    letter,
    slug: slugifyLetter(letter),
  }))

  return formattedLetters
}

export default {
  getAllWords,
  getByLetter,
  getWord,
  getAlphabet,
}
