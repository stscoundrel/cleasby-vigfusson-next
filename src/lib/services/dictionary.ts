import { getDictionary, DictionaryEntry as RawDictionaryEntry } from 'cleasby-vigfusson-dictionary'
import { VALID_AS_FIRST } from 'old-norse-alphabet'
import { oldNorseSort } from 'old-norse-alphabet-sort'
import { slugifyWord, slugifyLetter } from '../utils/slugs'

export interface DictionaryEntry extends RawDictionaryEntry {
  slug: string,
}

export interface DictionaryEntryDTO {
  word: string,
  slug: string
}

export interface AlphabetLetter {
  letter: string,
  slug: string
}

let cachedDictionary: DictionaryEntry[] | null = null

const addSlugs = (words: RawDictionaryEntry[]): DictionaryEntry[] => {
  const existingSlugs = {}

  const formattedWords = words.map((word) => {
    let slug = slugifyWord(word.word).toLowerCase()

    if (slug.charAt(0) === '-') {
      slug = slug.substring(1)
    }

    if (existingSlugs[slug]) {
      // Double slug, make unique.
      existingSlugs[slug] += 1
      slug = `${slug}-${existingSlugs[slug]}`
    } else {
      existingSlugs[slug] = 1
    }

    return {
      ...word,
      slug,
    }
  })

  return formattedWords
}

export const getAllWords = (): DictionaryEntry[] => {
  if (cachedDictionary) return cachedDictionary

  const words = getDictionary()

  /**
   * Add URL safe slugs.
   */
  const formattedWords = addSlugs(words)

  /**
   * Ensure headwords are lowercased.
   * Original source has some in all caps.
   * That is true to the layout of the book,
   * but makes little sense in website.
   */
  const dictionary = formattedWords.map((entry) => ({ ...entry, word: entry.word.toLowerCase() }))

  cachedDictionary = dictionary

  return dictionary
}

export const getByLetter = (letter: string): DictionaryEntryDTO[] => {
  const words = getAllWords()
  const byLetter = words
    .filter((entry) => (
      entry.word.charAt(0).toLowerCase() === letter.toLowerCase()))
    .map((entry) => {
      // Simpler DTO dictionary entry.
      const { word, slug } = entry
      return { word, slug }
    })
    .sort((a, b) => oldNorseSort(a.word, b.word))

  return byLetter
}

export const getWord = (slug: string): DictionaryEntry => (
  getAllWords().filter((entry) => entry.slug === slug)[0]
)

export const getRandomEntries = (): DictionaryEntry[] => (
  // Return entries fit to be randomized "teasers"
  // Therefore, content should be short, but not too short.
  getAllWords()
    .sort(() => Math.random() - 0.5)
    .filter((entry) => entry.definitions[0].length < 50 && entry.definitions[0].length > 15)
    .slice(0, 36)
    .sort((a, b) => oldNorseSort(a.word, b.word))
)

export const getSimilarWords = (entry: DictionaryEntry) : DictionaryEntry[] => getAllWords()
  .filter((dEntry) => dEntry.word.toLocaleLowerCase() === entry.word.toLocaleLowerCase()
  && dEntry.slug !== entry.slug)

export const getAlphabet = (): AlphabetLetter[] => {
  const letters = [...VALID_AS_FIRST.filter((letter) => letter !== 'ǫ' && letter !== 'ø'), 'ö']

  const formattedLetters = letters.map((letter) => ({
    letter,
    slug: slugifyLetter(letter),
  }))

  return formattedLetters
}

export default {
  getAllWords,
  getByLetter,
  getWord,
  getRandomEntries,
  getAlphabet,
}
