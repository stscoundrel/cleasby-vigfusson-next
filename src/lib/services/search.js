import { markWords } from 'markari'

const formatResults = (results, search, criteria) => {
  const formattedResults = results.map((result) => {
    const foundIn = []

    if (criteria.includes('definitions')) {
      result.definitions.forEach((definition) => {
        if (definition.toLowerCase().includes(search.toLowerCase())) {
          const highlighterDefinition = markWords(search, definition)
          foundIn.push(highlighterDefinition)
        }
      })
    }

    if (foundIn.length === 0) {
      const highlightedHeadword = markWords(search, result.word)
      foundIn.push(`In headword: ${highlightedHeadword}`)
    }

    return {
      ...result,
      foundIn,
    }
  })

  return formattedResults
}

export const searchDictionary = (search, dictionary, criteria = ['headword', 'definitions']) => {
  const filteredSearch = search.toLowerCase()

  const results = dictionary.filter((entry) => {
    let matchesSearch = false

    if (criteria.includes('headword')) {
      if (entry.word.toLowerCase().includes(filteredSearch)) {
        matchesSearch = true
      }

      if (entry.slug.includes(filteredSearch)) {
        matchesSearch = true
      }
    }

    if (criteria.includes('definitions')) {
      entry.definitions.forEach((definition) => {
        if (definition.toLowerCase().includes(filteredSearch)) {
          matchesSearch = true
        }
      })
    }

    return matchesSearch
  })

  const formattedResult = formatResults(results, search, criteria)

  return formattedResult
}

export default {
  searchDictionary,
}
