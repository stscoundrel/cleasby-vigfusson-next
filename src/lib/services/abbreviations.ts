import abbreviationsService from 'cleasby-vigfusson-abbreviations'
import { abbreviate } from 'abbreviatrix'
import { DictionaryEntry } from 'lib/services/dictionary'

const { findAbbreviations, findWorksAndAuthors, getWorksAndAuthorsMapping } = abbreviationsService

export interface Abbreviation{
  abbreviation: string,
  explanation: string
}

export interface CombinedAbbreviations {
  common: Abbreviation[],
  works: Abbreviation[],
}

/**
 * Combine abbreviations from definitions array.
 * Accepts finder method to use.
 */
const combineAbbreviations = (
  { definitions }: DictionaryEntry,
  abbreviationFinder: (s: string) => Map<string, string>,
): Abbreviation[] => {
  const combinedAbbreviations: Abbreviation[] = []
  const abbreviationSet = new Set()

  definitions.forEach((definition) => {
    const abbreviations = abbreviationFinder(definition)
    abbreviations.forEach((explanation, abbreviation) => {
      if (!abbreviationSet.has(abbreviation)) {
        abbreviationSet.add(abbreviation)
        combinedAbbreviations.push({ abbreviation, explanation })
      }
    })
  })

  return combinedAbbreviations;
}

export const getAbbreviations = (entry: DictionaryEntry): CombinedAbbreviations => ({
  common: combineAbbreviations(entry, findAbbreviations),
  works: combineAbbreviations(entry, findWorksAndAuthors),
})

export const getAllSorces = (): Abbreviation[] => {
  const abbrs = getWorksAndAuthorsMapping()
  return Array
    .from(abbrs, ([abbreviation, explanation]) => ({ explanation, abbreviation }))
    .sort((a, b) => a.abbreviation.localeCompare(b.abbreviation))
}

/**
 * Add abbr tags to content with explanations.
 */
export const addAbbreviationsToContent = (
  content: string,
  abbreviations: CombinedAbbreviations,
): string => {
  const combinedAbbrs = [...abbreviations.common, ...abbreviations.works]
  let result = content

  combinedAbbrs.forEach(({ abbreviation, explanation }) => {
    result = abbreviate(abbreviation, explanation, result)
  })

  return result
}

export default {
  getAbbreviations,
  addAbbreviationsToContent,
}
