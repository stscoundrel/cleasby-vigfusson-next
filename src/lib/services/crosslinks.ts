import { getCrosslinks as getAllCrossLinks, Crosslink, getOldNorseCrosslinks } from 'scandinavian-dictionary-crosslinker'
import { DictionaryEntry } from 'lib/services/dictionary'

export type { Crosslink } from 'scandinavian-dictionary-crosslinker'

let crossLinkCache: Record<string, Crosslink[]> | null = null

const populateCrossLinks = () => {
  if (!crossLinkCache) {
    crossLinkCache = getAllCrossLinks()
  }

  return crossLinkCache
}

export const getCrossLinks = (entry: DictionaryEntry): Crosslink[] => {
  const crosslinks = populateCrossLinks()
  if (Object.prototype.hasOwnProperty.call(crosslinks, entry.slug)) {
    return getOldNorseCrosslinks(entry.slug)
  }

  return []
}
