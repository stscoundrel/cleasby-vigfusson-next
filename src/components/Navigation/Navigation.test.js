import ReactDOM from 'react-dom/client'
import renderer from 'react-test-renderer'
import { getAlphabet } from 'lib/services/dictionary'
import Navigation from './index'

const letters = getAlphabet()

describe('Navigation component', () => {
  test('Does not crash', () => {
    const div = document.createElement('div')
    const root = ReactDOM.createRoot(div)
    root.render(<Navigation letters={letters} />)
  })

  test('Matches snapshot', () => {
    const tree = renderer.create(<Navigation letters={letters} />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
