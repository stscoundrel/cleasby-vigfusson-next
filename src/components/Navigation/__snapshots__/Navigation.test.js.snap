// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation component Matches snapshot 1`] = `
Array [
  <div
    className="topBarSpacer"
  />,
  <nav
    className="section"
  >
    <div
      className="topbar container"
    >
      <img
        alt="To home"
        height="30"
        loading="lazy"
        onClick={[Function]}
        onMouseEnter={[Function]}
        src="/favicon-48x48.png"
        width="30"
      />
      <div
        aria-label="Open menu"
        className="hamburger"
        onClick={[Function]}
        role="button"
      >
        <span
          className="part"
        />
        <span
          className="part"
        />
        <span
          className="part"
        />
      </div>
    </div>
    <div
      className="content null container"
    >
      <ul
        className="list"
      >
        <li
          className="listItem"
        >
          <a
            className="link"
            href="/letter/a"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            a
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            className="link"
            href="/letter/a2"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            á
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            className="link"
            href="/letter/b"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            b
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            className="link"
            href="/letter/d"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            d
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            className="link"
            href="/letter/e"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            e
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            className="link"
            href="/letter/e2"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            é
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            className="link"
            href="/letter/f"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            f
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            className="link"
            href="/letter/g"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            g
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            className="link"
            href="/letter/h"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            h
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            className="link"
            href="/letter/i"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            i
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            className="link"
            href="/letter/i2"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            í
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            className="link"
            href="/letter/j"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            j
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            className="link"
            href="/letter/k"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            k
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            className="link"
            href="/letter/l"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            l
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            className="link"
            href="/letter/m"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            m
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            className="link"
            href="/letter/n"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            n
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            className="link"
            href="/letter/o"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            o
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            className="link"
            href="/letter/o2"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            ó
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            className="link"
            href="/letter/p"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            p
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            className="link"
            href="/letter/r"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            r
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            className="link"
            href="/letter/s"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            s
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            className="link"
            href="/letter/t"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            t
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            className="link"
            href="/letter/u"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            u
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            className="link"
            href="/letter/u2"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            ú
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            className="link"
            href="/letter/v"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            v
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            className="link"
            href="/letter/y"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            y
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            className="link"
            href="/letter/y2"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            ý
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            className="link"
            href="/letter/th"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            þ
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            className="link"
            href="/letter/ae"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            æ
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            className="link"
            href="/letter/oe3"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            œ
          </a>
        </li>
        <li
          className="listItem"
        >
          <a
            className="link"
            href="/letter/oe"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            ö
          </a>
        </li>
      </ul>
      <form
        className="form"
        onSubmit={[Function]}
      >
        <input
          aria-label="Search"
          className="input"
          name="search"
          onChange={[Function]}
          type="search"
          value=""
        />
        <button
          className="button button--secondary button--no-margs"
          name="submit"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
  </nav>,
]
`;
